design/capm3-remediation-controller-proposal.md

```diff
@@ -0,0 +1,382 @@
+
+## Motivation
+
+It is not always practical to require admin intervention once a node has been
+identified as having reached a bad or unknown state. In order to automate the
+recovery of unhealthy nodes we need a programmatic way to put them back into a
+safe and healthy state.
+
+The Cluster API includes an optional
+[Machine Healthcheck](https://cluster-api.sigs.k8s.io/tasks/healthcheck.html)
+component that implements automated health checking capability, and the with
+[External Remediation proposal](https://github.com/kubernetes-sigs/cluster-api/pull/3190)
+it will be possible to plug in Metal3 specific remediation strategies to
+remediate an unhealthy node while relying on Cluster
+API MHC to determine unhealthy nodes.
+
+We would like to enable an automated mechanism in Metal3 that supports a
+recovery flow tailored for bare metal environments whenever Machine(s)
+meet unhealthiness criteria set by Cluster API MHC.
+
+### Goals
+
+* Enable automated remediation that doesn’t always reprovision the node.
+* Integrate with Cluster API MHC.
+* Support remediation based on power cycling the underlying hardware.
+* Support use of BMO reboot, power on/off APIs and
+  CAPM3 unhealthy annotation as part of automated remediation cycles.
+
+### Non-Goals
+
+* Create a CAPM3 specific Machine Healthcheck.
+* Adding new remediation or power cycle APIs into BMO.
+
+## Proposal
+
+This proposal calls for a new Metal3Remediation object and a new controller in
+CAPM3 that watches them and uses BMH APIs to perform automated remediation of
+unhealthy nodes in Metal3 project.
+
+The controller will locate the Machine with the same name as the
+Metal3Remediation CR and uses existing BMO and CAPM3 apis to remediate
+associated unhealthy baremetal nodes.
+
+New Remediation controller supports Reboot Strategy specified in
+Metal3Remediation CRD and the new controller uses the same object to store
+the state of the current remediation cycle.
+
+This proposal also introduces a Metal3RemediationTemplate CRD. The
+Metal3RemediationTemplate is used by CAPI MachineHealthCheck to create
+Metal3Remediation objects when it detects unhealthy Machines in the cluster.
+
+### User Stories
+
+#### Story 1
+
+As an administrator, I would like my cluster to be highly available and nodes
+be self healing in case they fail to operate.
+
+#### Story 2
+
+As an administrator, I would like to have a simple way of requesting recovery
+for my node when I find it to be unhealthy.
+
+#### Story 3
+
+As an admin, I would like to view historical data about the health of my
+cluster nodes for debugging purposes.
+
+#### Story 4
+
+As an admin, I should be able to mark ready nodes to be unusable.
+
+### Design Details
+
+CAPI machine health check allows the infrastructure provider to implement its
+own remediation controller to reconcile External Machine Remediation (EMR) CRs.
+EMRs are generated using infrastructure provider defined External Remediation
+Template.
```

Comment:
So an External Machine Remediation resource isn't a "type" defined by CAPI, it uses the template defined by the provider?

Reply:
Yes, that is correct.